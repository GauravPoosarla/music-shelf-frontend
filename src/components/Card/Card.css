.card {
  display: flex;
  flex-direction: column;
  height: 200px;
  width: 200px;
  margin: 10px;
  border: 1px solid #e0e0e0;
  border-radius: 10px;
  box-shadow: 0 0 5px 0 rgba(0, 0, 0, 0.2);
  padding: 10px
}

.card-light {
  background-color: #dfdfdf;
}

.card-dark {
  background-color: #ababab;
}

.card img {
  height: 120px;
  width: 120px;
  border-radius: 50%;
  box-sizing: border-box;
  margin-top: 10px;
  margin-left: 40px;
  border: 1px solid black;
}

.song-user {
  display: flex;
  flex-direction: row;
  margin-top: 20px;
  height: 100px;
  justify-content: space-between;
}

.song-info {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
}

.song-name {
  margin-top: 5px;
  font-size: 15px;
  height: 20px;
  overflow: hidden;
}

.song-artist {
  font-size: 10px;
  height: 20px;
  font-weight: bold;
  overflow: hidden;
  text-overflow: ellipsis;
}

.like img{
  height: 25px;
  width: 25px;
  border-radius: 0%;
  border: none;
}

.like p {
  color: white;
  font-size: 10px;
  margin-top: -25px;
  margin-left: 40px;
}

.btn-like {
  background-color: transparent;
  border: none;
  color: white;
  width: 30px;
  height: 30px;
  background-color: transparent;
}


@media screen and (max-width: 768px) {
  .card {
    height: 250px;
    width: 300px;
  }

  .card img {
    height: 180px;
    width: 180px;
    margin-left: 60px;
  }

  .song-user {
    height: 50px;
    margin: 0px;
  }

  .song-name {
    height: 30px;
    font-size: 20px;
  }

  .song-artist {
    font-size: 15px;
  }
  
  .like img{
    height: 20px;
    width: 20px;
  }
  
  .like p {
    margin-top: -23px;
    margin-left: 50px;
  }
}
  
@media screen and (max-width: 400px) {
  .card {
    height: 150px;
    width: 200px;
  }

  .card img {
    height: 100px;
    width: 100px;
    margin-left: 50px;
    margin-top: 5px;
  }

  .song-user {
    height: 50px;
    margin: 0px;
  }

  .song-name {
    height: 30px;
    font-size: 15px;
  }

  .song-artist {
    height: 20px;
    font-size: 10px;
  }
  
  .like img{
    height: 15px;
    width: 15px;
  }
  
  .like p {
    margin-top: -20px;
    margin-left: 40px;
  }
}