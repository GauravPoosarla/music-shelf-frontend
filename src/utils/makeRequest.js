import axios from 'axios';
import { BACKEND_URL } from '../constants/apiEndPoint';

const makeRequest = async ({ url, method }, dynamicConfig, navigate) => {
  try {
    const response = await axios({
      baseURL: BACKEND_URL,
      url,
      method,
      ...dynamicConfig,
    });

    return response.data;
  } catch (error) {
    console.log(error);
    if (error.response.status === 401) {
      navigate('/error');
    }
  }
};

export default makeRequest;
