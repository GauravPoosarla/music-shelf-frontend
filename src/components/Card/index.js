import React from 'react';
import './Card.css';
import { GET_SONG_DATA } from '../../constants/apiEndPoint';
import makeRequest from '../../utils/makeRequest';
import heartRed from '../../assets/heart-red.svg';
import { useState } from 'react';

function Card({ song }) {
  const [isLiked, setIsLiked] = useState(false);

  return (
    <div className="card">
      <img src={song.imageUrl} alt="song img" />
      <div className="song-user">
        <div className="song-info">
          <p className="song-name">{song.name}</p>
          <p className="song-artist">{song.artist.name}</p>
        </div>
        <div className="like">
          <img src={heartRed} alt="heart" />
        </div>
      </div>
    </div>
  );
}

export default Card;
