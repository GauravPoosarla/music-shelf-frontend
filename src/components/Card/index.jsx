import React from 'react';
import './Card.css';
import { GET_SONG_DATA, LIKE_SONG } from '../../constants/apiEndPoint';
import makeRequest from '../../utils/makeRequest';
import heartRed from '../../assets/heart-red.svg';
import heartGray from '../../assets/heart-gray.svg';
import { useState, useEffect } from 'react';

function Card(props) {
  const { song, index } = props;
  const [isLiked, setIsLiked] = useState(false);
  const [count, setCount] = useState(0);

  useEffect(() => {
    makeRequest(GET_SONG_DATA(song.id), {
      headers: { authorization: 'Bearer QWlzaHdhcnlhIE4=' },
    }).then((data) => {
      setIsLiked(data.data.like);
      setCount(data.data.count);
    });
  }, []);

  const likeHandler = () => {
    if (isLiked) {
      makeRequest(LIKE_SONG(song.id), {
        headers: { authorization: 'Bearer QWlzaHdhcnlhIE4=' },
        data: { like: false },
      });
      setIsLiked(!isLiked);
      setCount(count - 1);
    } else {
      makeRequest(LIKE_SONG(song.id), {
        headers: { Authorization: 'Bearer QWlzaHdhcnlhIE4=' },
        data: { like: true },
      });
      setIsLiked(!isLiked);
      setCount(count + 1);
    }
  };

  const color = index % 2 === 0 ? 'dark' : 'light';

  return (
    <div className={`card card-${color}`}>
      <img src={song.imageUrl} alt="song img" />
      <div className="song-user">
        <div className="song-info">
          <p className="song-name">{song.name}</p>
          <p className="song-artist">{song.artist.name}</p>
        </div>
        <div className="like">
          {isLiked ? (
            <div>
              <button
                className="btn-like"
                style={{
                  background: `url(${heartRed}) no-repeat center / contain`,
                }}
                onClick={likeHandler}
              >
                {count}
              </button>
            </div>
          ) : (
            <div>
              <button
                className="btn-like"
                style={{
                  background: `url(${heartGray}) no-repeat center / contain`,
                }}
                onClick={likeHandler}
              >
                {count}
              </button>
            </div>
          )}
        </div>
      </div>
    </div>
  );
}

export default Card;
