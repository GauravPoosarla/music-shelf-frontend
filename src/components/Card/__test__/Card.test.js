import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';

import Card from '..';
import makeRequest from '../../../utils/makeRequest';

jest.mock('../../../utils/makeRequest');

describe('Card', () => {
  const mockSong = {
    id: 'cd3cc1e3-e1e0-4ccd-bc67-747648985838',
    name: 'Lost',
    imageUrl:
      'https://i.scdn.co/image/ab67616d0000b27386a8ab515de4b7aef28cd631',
    artist: {
      id: '496b0a85-2bfa-45bc-8d0f-57fe0ce55708',
      name: 'Maroon 5',
    },
    genre: {
      id: '128aa7f8-c943-48ce-b352-7edd26fa4c6e',
      name: 'Pop',
    },
  };

  it('should render the card', () => {
    makeRequest.mockResolvedValue({
      data: {
        count: 15,
        like: true,
      },
    });
    render(<Card song={mockSong} />);
    expect(screen.queryByText('Lost')).toBeTruthy();
    waitFor(() => {
      expect(screen.queryByText('15')).toBeTruthy();
    });
  });

  it('should render the card with like false', () => {
    makeRequest.mockResolvedValueOnce({
      data: {
        count: 15,
        like: false,
      },
    });
    makeRequest.mockResolvedValueOnce({
      data: {
        count: 16,
        like: true,
      },
    });
    const { container } = render(<Card song={mockSong} />);
    waitFor(() => {
      expect(screen.queryByText('15')).toBeTruthy();
    });

    const likeButton = container.querySelector('.like');
    fireEvent.click(likeButton);

    waitFor(() => {
      expect(screen.queryByText('16')).toBeTruthy();
    });
  });

  it('should toggle the like image when clicked', () => {
    makeRequest.mockResolvedValueOnce({
      data: {
        count: 15,
        like: false,
      },
    });
    makeRequest.mockResolvedValueOnce({
      data: {
        count: 16,
        like: true,
      },
    });

    const { container } = render(<Card song={mockSong} />);
    const likeButton = container.querySelector('.like');

    waitFor(() => {
      expect(likeButton.src).toContain('like-gray.svg');
    });

    fireEvent.click(likeButton);

    waitFor(() => {
      expect(likeButton.src).toContain('like-red.svg');
    });

    fireEvent.click(likeButton);

    waitFor(() => {
      expect(likeButton.src).toContain('like-gray.svg');
    });
  });

  it('should render correctly', () => {
    makeRequest.mockResolvedValue({
      data: {
        count: 15,
        like: true,
      },
    });
    const { container } = render(<Card song={mockSong} />);
    expect(container).toMatchSnapshot();
  });
});
