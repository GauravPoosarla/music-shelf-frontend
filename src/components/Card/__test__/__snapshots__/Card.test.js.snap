// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render correctly 1`] = `
<div>
  <div
    class="card card-light"
  >
    <img
      alt="song img"
      src="https://i.scdn.co/image/ab67616d0000b27386a8ab515de4b7aef28cd631"
    />
    <div
      class="song-user"
    >
      <div
        class="song-info"
      >
        <p
          class="song-name"
        >
          Lost
        </p>
        <p
          class="song-artist"
        >
          Maroon 5
        </p>
      </div>
      <div
        class="like"
      >
        <div>
          <button
            class="btn-like"
          >
            0
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
